[Suggest] Added selectedItem prop

danilotorrisi opened the pull request:
#### Fixes #2784

#### Checklist
<!-- fill this section out if necessary, remove it otherwise -->

- [x] [Enable CircleCI for your fork](https://circleci.com/add-projects)
- [ ] Include tests
- [ ] Update documentation

#### Changes proposed in this pull request:

Fixes #2784.

#### Reviewers should focus on:

The `Suggest` component now takes an optional selectedItem prop. If defined the selectedItem state will be initialised with the given selectedItem.

palantirtech commented:
Thanks for your interest in [palantir/blueprint](https://github.com/palantir/blueprint), @danilotorrisi! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

danilotorrisi commented:
<h3>Init the Suggest state with the selected item prop.</h3>

Preview: <strong><a href='https://31-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/docs-app/dist/index.html' target='_blank'>documentation</a> | <a href='https://31-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/landing-app/dist/index.html' target='_blank'>landing</a> | <a href='https://31-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/table-dev-app/dist/index.html' target='_blank'>table</a></strong>

giladgray commented on packages/select/src/components/select/suggest.tsx:
```diff
@@ -68,6 +74,7 @@ export class Suggest<T> extends React.PureComponent<ISuggestProps<T>, ISuggestSt
 
     public state: ISuggestState<T> = {
         isOpen: (this.props.popoverProps && this.props.popoverProps.isOpen) || false,
+        selectedItem: this.props.selectedItem
```
what you've implemented here is `defaultSelectedItem` - the initial uncontrolled state.

`selectedItem` should _control_ the state field by updating it when props change: if i pass a new `selectedItem` prop value, that should be reflected in later renders. this will not happen with your current code.

giladgray replied:
this is analogous to `value`, `defaultValue` and `onChange` on an `<input>` tag. 
https://reactjs.org/docs/uncontrolled-components.html

danilotorrisi replied:
You're right, I made the change to fix one bug that I encountered on my code but I missed that once the `selectedItem` has been initialised, the later changes won't do anything at all. I'll make a change right away.

danilotorrisi replied:
@giladgray In controlled mode, do you think the Suggest should still set the `selectedItem` state in the `handleItemSelect` handler? The `setState` will be called anyway because of the `nextOpenState` change, otherwise we could save a render cycle.

danilotorrisi commented:
<h3>Fixed a tslint coma issue.</h3>

Preview: <strong><a href='https://45-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/docs-app/dist/index.html' target='_blank'>documentation</a> | <a href='https://45-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/landing-app/dist/index.html' target='_blank'>landing</a> | <a href='https://45-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/table-dev-app/dist/index.html' target='_blank'>table</a></strong>

danilotorrisi commented:
<h3>Fixed Suggest support for controlled mode.</h3>

Preview: <strong><a href='https://44-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/docs-app/dist/index.html' target='_blank'>documentation</a> | <a href='https://44-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/landing-app/dist/index.html' target='_blank'>landing</a> | <a href='https://44-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/table-dev-app/dist/index.html' target='_blank'>table</a></strong>

giladgray commented on packages/select/src/components/select/suggest.tsx:
```diff
@@ -104,10 +110,12 @@ export class Suggest<T> extends React.PureComponent<ISuggestProps<T>, ISuggestSt
         const { placeholder = "Search..." } = inputProps;
 
+        // Let's support both controlled and uncontrolled.
+        const selectedItem = this.props.selectedItem || this.state.selectedItem;
```
this is not how i've implemented un/controlled behavior in the past. i prefer to use `this.state` as the single point of truth and to put `props` changes into state using lifecycle methods.

giladgray replied:
also please add unit tests for this behavior.

danilotorrisi commented:
@giladgray  I already use TSLint for my projects, but for some reason the integration with Code was not working as expected. I changed the implementation using the state as the only source of truth, also added some tests.

danilotorrisi commented:
<h3>Made the state the only source of truth, added tests.</h3>

Preview: <strong><a href='https://52-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/docs-app/dist/index.html' target='_blank'>documentation</a> | <a href='https://52-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/landing-app/dist/index.html' target='_blank'>landing</a> | <a href='https://52-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/table-dev-app/dist/index.html' target='_blank'>table</a></strong>

danilotorrisi commented on packages/select/src/components/select/suggest.tsx:
```diff
@@ -102,6 +103,13 @@ export class Suggest<T> extends React.PureComponent<ISuggestProps<T>, ISuggestSt
 
+    public componentWillReceiveProps(nextProps: ISuggestProps<T>) {
+        // If the selected item prop changes, update the underlying state.
+        if (nextProps.selectedItem !== this.state.selectedItem) {
```
Maybe it should be better to use isEqual from lodash instead of the simple operator, what do you think?

giladgray replied:
❌ not adding a dependency on lodash. this issue is common to all the select components and i think should be solved by an optional `itemEquals` prop a la `itemDisabled`

giladgray replied:
but not in this PR 👍

giladgray commented on packages/select/test/suggestTests.tsx:
```diff
@@ -10,6 +10,7 @@ import { mount, ReactWrapper } from "enzyme";
 import * as React from "react";
 import * as sinon from "sinon";
 
+import { first } from "lodash";
```
we don't have a dependency on lodash. please just use `[0]`.

giladgray commented on packages/select/test/suggestTests.tsx:
```diff
@@ -210,6 +211,23 @@ describe("Suggest", () => {
         }
     });
 
+    describe("Controlled Mode", () => {
```
please also test that interactions do not get into a weird intermediate state. see other test suites.

danilotorrisi commented:
<h3>Added more tests, improved controlled mode support</h3>

Preview: <strong><a href='https://59-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/docs-app/dist/index.html' target='_blank'>documentation</a> | <a href='https://59-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/landing-app/dist/index.html' target='_blank'>landing</a> | <a href='https://59-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/table-dev-app/dist/index.html' target='_blank'>table</a></strong>

giladgray commented on packages/select/src/components/select/suggest.tsx:
```diff
@@ -188,11 +188,16 @@ export class Suggest<T> extends React.PureComponent<ISuggestProps<T>, ISuggestSt
-            isOpen: nextOpenState,
-            selectedItem: item,
-        });
+        // just like EditableText, the internal state should only change when uncontrolled.
```
no need to reference `EditableText`, this is a very common React pattern

giladgray commented on packages/select/src/components/select/suggest.tsx:
```diff
@@ -188,11 +188,16 @@ export class Suggest<T> extends React.PureComponent<ISuggestProps<T>, ISuggestSt
-            selectedItem: item,
-        });
+        // just like EditableText, the internal state should only change when uncontrolled.
+        if (this.props.selectedItem == null) {
```
i think you'll want `=== undefined` here to allow `null` to mean `no selection`

giladgray commented on packages/select/src/components/select/suggest.tsx:
```diff
@@ -38,6 +38,12 @@ export interface ISuggestProps<T> extends IListItemsProps<T> {
+     * Optionally, if provided the selected item will be in controlled mode.
+     * Use the onItemSelect function to monitor changes.
+     */
+    selectedItem?: T;
```
make this `T | null` so `null` can indicate "controlled empty selection"

danilotorrisi replied:
Now it supports the controlled empty selection. I've also added a tests for this specific scenario.

danilotorrisi replied:
@giladgray does it look to you?

danilotorrisi commented:
<h3>Added support for controlled empty selection.</h3>

Preview: <strong><a href='https://66-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/docs-app/dist/index.html' target='_blank'>documentation</a> | <a href='https://66-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/landing-app/dist/index.html' target='_blank'>landing</a> | <a href='https://66-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/table-dev-app/dist/index.html' target='_blank'>table</a></strong>

switz commented:
This would be very helpful -- does anything need to be changed before a merge & deploy?

giladgray commented on packages/select/src/components/select/suggest.tsx:
```diff
@@ -38,6 +38,12 @@ export interface ISuggestProps<T> extends IListItemsProps<T> {
 
+    /**
+     * Optionally, if provided the selected item will be in controlled mode.
+     * Use the onItemSelect function to monitor changes.
```
please edit this to match `activeItem` docs (including backticks around prop name): 

> The currently selected item, or null to indicate that no item is selected. If omitted, this prop will be uncontrolled (managed by the component's state). Use `onItemSelect` to listen for updates.

danilotorrisi commented:
<h3>prop docs</h3>

Preview: <strong><a href='https://79-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/docs-app/dist/index.html' target='_blank'>documentation</a> | <a href='https://79-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/landing-app/dist/index.html' target='_blank'>landing</a> | <a href='https://79-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/table-dev-app/dist/index.html' target='_blank'>table</a></strong>

danilotorrisi commented:
<h3>whitespace</h3>

Preview: <strong><a href='https://80-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/docs-app/dist/index.html' target='_blank'>documentation</a> | <a href='https://80-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/landing-app/dist/index.html' target='_blank'>landing</a> | <a href='https://80-146469272-gh.circle-artifacts.com/0/home/circleci/project/packages/table-dev-app/dist/index.html' target='_blank'>table</a></strong>
